src/lstm_layer.cpp

```diff
@@ -1515,3 +1515,505 @@ void LSTM::preinit_layer() {
+        for (int j = 0; j < no; j++) {
+            // Forget gate
+            Czz_f = var_h_prev[t] * Jca[j] * mo_ga[j] * Jf_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_f] * mc_prev[j];
+
+            // Input gate
+            Czz_i = var_h_prev[t] * Jca[j] * mo_ga[j] * Ji_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_i] * mc_ga[j];
+
+            // Cell state gate
+            Czz_c = var_h_prev[t] * Jca[j] * mo_ga[j] * Jc_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_c] * mi_ga[j];
+
+            // Output gate
+            Czz_o = var_h_prev[t] * Jo_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_o] * mca[j];
+
+            // Updating quantities
+            m = t * no + j;
+            Ch_h[m] = Czz_f + Czz_i + Czz_c + Czz_o;
+        }
+    }
+    cov_hh.push_back(Ch_h);
+}
+
+void lstm_smoother_cell_states(int num_timestep,
+                               std::vector<std::vector<float>> &cov_cc,
+                               std::vector<std::vector<float>> &mu_c_priors,
+                               std::vector<std::vector<float>> &var_c_priors,
+                               std::vector<std::vector<float>> &mu_c_posts,
+                               std::vector<std::vector<float>> &var_c_posts,
+                               std::vector<std::vector<float>> &mu_c_smooths,
+                               std::vector<std::vector<float>> &var_c_smooths)
+/*
+ */
+{
+    int num_states = mu_c_priors[0].size();
+    for (int i = num_timestep - 2; i >= 0; --i) {
+        for (int j = 0; j < num_states; j++) {
+            float tmp = cov_cc[i + 1][j] / var_c_priors[i + 1][j];
+            mu_c_smooths[i][j] =
+                mu_c_posts[i][j] +
+                tmp * (mu_c_smooths[i + 1][j] - mu_c_priors[i + 1][j]);
+            var_c_smooths[i][j] =
+                var_c_posts[i][j] +
+                tmp * (var_c_smooths[i + 1][j] - var_c_priors[i + 1][j]) * tmp;
+        }
+    }
+}
+
+void lstm_smoother_hidden_states(int num_timestep,
+                                 std::vector<std::vector<float>> &cov_hc,
+                                 std::vector<std::vector<float>> &mu_c_priors,
+                                 std::vector<std::vector<float>> &var_c_priors,
+                                 std::vector<std::vector<float>> &mu_h_posts,
+                                 std::vector<std::vector<float>> &var_h_posts,
+                                 std::vector<std::vector<float>> &mu_c_smooths,
+                                 std::vector<std::vector<float>> &var_c_smooths,
+                                 std::vector<std::vector<float>> &mu_h_smooths,
+                                 std::vector<std::vector<float>> &var_h_smooths)
+/*
+ */
+{
+    int num_states = mu_c_priors[0].size();
+    for (int i = num_timestep - 2; i >= 0; i--) {
+        for (int j = 0; j < num_states; j++) {
+            float tmp = cov_hc[i + 1][j] / var_c_priors[i + 1][j];
```

Comment:
I would prefer to use 1D vector as a future cuda vector. A matrix M[nxm] where n is a leading dimension M[i][j] = M[i * m + j]